pkg/mimo/actuator/manager.go

```diff
@@ -0,0 +1,182 @@
+package actuator
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/maps"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/mimo/tasks"
+)
+
+const maxDequeueCount = 5
+
+type Actuator interface {
+	Process(context.Context) (bool, error)
+	AddTask(string, tasks.TaskFunc)
+	AddTasks(map[string]tasks.TaskFunc)
+}
+
+type actuator struct {
+	env env.Interface
+	log *logrus.Entry
+	now func() time.Time
+
+	clusterID string
+
+	oc  database.OpenShiftClusters
+	mmf database.MaintenanceManifests
+
+	tasks map[string]tasks.TaskFunc
+}
+
+func NewActuator(
+	ctx context.Context,
+	_env env.Interface,
+	log *logrus.Entry,
+	clusterID string,
+	oc database.OpenShiftClusters,
+	mmf database.MaintenanceManifests,
+	now func() time.Time) (Actuator, error) {
+	a := &actuator{
+		env:       _env,
+		log:       log,
+		clusterID: strings.ToLower(clusterID),
+		oc:        oc,
+		mmf:       mmf,
+		tasks:     make(map[string]tasks.TaskFunc),
+
+		now: now,
+	}
+
+	return a, nil
+}
+
+func (a *actuator) AddTask(u string, t tasks.TaskFunc) {
+	a.tasks[u] = t
+}
+
+func (a *actuator) AddTasks(tasks map[string]tasks.TaskFunc) {
+	maps.Copy(a.tasks, tasks)
+}
+
+func (a *actuator) Process(ctx context.Context) (bool, error) {
+	// Get the manifests for this cluster which need to be worked
+	i, err := a.mmf.GetByClusterID(ctx, a.clusterID, "")
+	if err != nil {
+		return false, err
+	}
+
+	docList := make([]*api.MaintenanceManifestDocument, 0)
+	for {
+		docs, err := i.Next(ctx, -1)
+		if err != nil {
+			return false, err
+		}
+		if docs == nil {
+			break
+		}
+
+		docList = append(docList, docs.MaintenanceManifestDocuments...)
+	}
+
+	manifestsToAction := make([]*api.MaintenanceManifestDocument, 0)
+
+	sort.SliceStable(docList, func(i, j int) bool {
+		if docList[i].MaintenanceManifest.RunAfter != docList[j].MaintenanceManifest.RunAfter {
+			return docList[i].MaintenanceManifest.Priority < docList[j].MaintenanceManifest.Priority
+		}
+
+		return docList[i].MaintenanceManifest.RunAfter < docList[j].MaintenanceManifest.RunAfter
+	})
+
+	evaluationTime := a.now()
+
+	// Check for manifests that have timed out first
+	for _, doc := range docList {
+		if evaluationTime.After(time.Unix(int64(doc.MaintenanceManifest.RunBefore), 0)) {
+			// timed out, mark as such
+			a.log.Infof("marking %v as outdated: %v older than %v", doc.ID, doc.MaintenanceManifest.RunBefore, evaluationTime.UTC())
+
+			_, err := a.mmf.Patch(ctx, a.clusterID, doc.ID, func(d *api.MaintenanceManifestDocument) error {
```

Comment:
shall we implement a retry logic here? just to make the patch action more robust?